src/pieces/pawn.cpp

```diff
@@ -0,0 +1,204 @@
+    if(test_coordinate.is_valid())
+    {
+        test_piece = board.get_piece_at(test_coordinate);
+        if(test_piece == nullptr)
+        {
+            return test_coordinate;
+        }
+    }
+    return this->coordinate; 
+}
+/************************** double step *****************************/
+Chess::Coordinate PawnPiece::double_step(Board& board,Color color)
+{
+    Coordinate test_coordinate = this->coordinate;
+    std::shared_ptr<Piece> test_piece;
+    if(!get_had_moved())
+    {
+        test_coordinate = one_step(board,one_step(board,test_coordinate,color),color);
+        return test_coordinate;
+    }
+    return test_coordinate;
+}
+
+/************************** diagonally step *****************************/
+std::pair<Chess::Coordinate,Chess::Coordinate> PawnPiece::diag_cap(Board& board, Color color)
+{
+    Coordinate test_coordinate;
+    std::shared_ptr<Piece> test_piece;
+    std::pair<Coordinate,Coordinate> valid_coordinate;
+    Direction direction1 = (color == Color::white ? Direction::left_up : Direction::left_down);
+    Direction direction2 = (color == Color::white ? Direction::right_up : Direction::right_down);
+    test_coordinate = this->coordinate + DirectionOffset.at(direction1);
+    if(test_coordinate.is_valid())
+    {
+        test_piece = board.get_piece_at(test_coordinate);
+        if(test_piece != nullptr && this->color != test_piece->get_color() )
+        {
+            valid_coordinate.first = test_coordinate;
+        }
+    }
+    test_coordinate = this->coordinate + DirectionOffset.at(direction2);
+    if(test_coordinate.is_valid())
+    {
+        test_piece = board.get_piece_at(test_coordinate);
+        if(test_piece != nullptr && this->color != test_piece->get_color() )
+        {
+            valid_coordinate.second = test_coordinate;
+        }
+    }
+    return valid_coordinate;
+}
+
+/************************** en passant (!)*****************************/
+Chess::Coordinate PawnPiece::en_passant(Board& board,Color color)
+{
+    Coordinate end_p, start_p;
+    std::shared_ptr<Piece> test_piece;
+    Direction direction = (color == Color::white ? Direction::up : Direction::down);
+    Direction direction_l = (color == Color::white ? Direction::left_up : Direction::left_down);
+    Direction direction_r = (color == Color::white ? Direction::right_up : Direction::right_down);
+    end_p = this->coordinate + DirectionOffset.at(Direction::left);
+    start_p = end_p + DirectionOffset.at(direction) + DirectionOffset.at(direction);
+    if(end_p.is_valid() && start_p.is_valid())
+    {
+        test_piece = board.get_piece_at(end_p);
+        if(test_piece != nullptr && this->color != test_piece->get_color() && this->type == test_piece->get_type())
+        {
+            if(board.get_last_movement().start == start_p && board.get_last_movement().end == end_p)
+            {
+                end_p = this->coordinate + DirectionOffset.at(direction_l);
```

Comment:
controlla se in end_p non ci siano pezzi dello stesso colore del pedone che ci vuole andare

Reply:
per il resto ok, puoi fare il merge subito dopo

Reply:
scherzavo ripensandoci non può esserci mai la possibilità che dietro al pedone avversario ci siano altri pezzi, perchè il pedone appunto si è appena mosso di due. Quindi ritorna la vecchia versione